sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala

```diff
@@ -222,12 +222,14 @@ class StreamingQueryManager private[sql] (sparkSession: SparkSession) extends Lo
     }.getOrElse {
       if (useTempCheckpointLocation) {
         deleteCheckpointOnStop = true
-        val tempDir = Utils.createTempDir(namePrefix = s"temporary").getCanonicalPath
+        val tempDir = System.getProperty("java.io.tmpdir")
+        val cpTempDir = new Path("file://" + tempDir + "/temporary-"
+          + UUID.randomUUID.toString).toString
```

Comment:
I've intentionally not used `toUri` here because of [#23733](https://github.com/apache/spark/pull/23733).

Reply:
I dont think this handles cleaning up the tmp directory on shutdown which is already handled in `Utils.createTempDir`. I assume here we want the checkpoint location to be on the local file system and not in HDFS, so why not just re-use the Utils.createTempDir and just prefix "file://" to the canonical path to fix the issues with resolution.

Reply:
I tend to agree with @arunmahadevan for leveraging `Utils.createTempDir`, but `file://` should be added as prefix of `root` parameter to `Utils.createTempDir` to prevent choosing default schema of filesystem and finally shutdown hook to delete directory from wrong filesystem.

Reply:
I am not sure adding "file://" prefix to root parameter would cause the canonical path to return the final path with the same prefix. The attempted resolution with HDFS Filesystem happens later so we may be ok to just prepend the "file://" to canonical path.

Reply:
> I am not sure adding "file://" prefix to root parameter would cause the canonical path to return the final path with the same prefix.

That actually sounds serious to me and even sounds one reason to avoid.

Reply:
I'm not sure why `Utils.createTempDir` doesn't work here? What is the problem with `mkdir` being called twice?

Also I'm not sure I understand why you're not using `.toUri`. Even ignoring the PR you reference, which is closed, won't the string here look exactly the same?

```
scala> val f = new java.io.File("/tmp")
f: java.io.File = /tmp

scala> new Path(new Path(f.toURI()), "foo").toString
res1: String = file:/tmp/foo

scala> new Path(new Path(f.toURI()), "foo").toUri.toString
res2: String = file:/tmp/foo
```

Also the "file:" scheme doesn't generally like "//". Which is why you should use `File.toURI` instead of writing your own...

Reply:
> I'm not sure why Utils.createTempDir doesn't work here? What is the problem with mkdir being called twice?

`Utils.createTempDir` would work but an optional fix not to do it twice. The main problem is filesystem changes all of a sudden when default fs is not local. Please see the added test.

> Also I'm not sure I understand why you're not using `.toUri`.

```
scala> import org.apache.hadoop.fs.Path
import org.apache.hadoop.fs.Path

scala> val f = new java.io.File("/tmp/chk %#chk")
f: java.io.File = /tmp/chk %#chk

scala> new Path(new Path(f.toURI()), "foo").toString
res0: String = file:/tmp/chk %#chk/foo

scala> new Path(new Path(f.toURI()), "foo").toUri.toString
res1: String = file:/tmp/chk%20%25%23chk/foo
```

I know it can happen rarely but since it's discovered and we know it's better to prepare the code.

Reply:
> Also the "file:" scheme doesn't generally like "//".

Not sure what you mean here (agree that manual URI creation is not the most appropriate way but considering the mentioned URI resolution this seemed the most reasonable way).

```
scala> val cpTempDir = new Path("file://" + tempDir + "/temporary1")
cpTempDir: org.apache.hadoop.fs.Path = file:/var/folders/t_/w90m85fn2gjb1v0c24wrfrxm0000gp/T/temporary1
```

Reply:
There have been bugs in the past caused by the host part (the "//") being included in file URIs. It's generally better to not add it since I don't believe it's even part of the respective RFC (although I didn't bother to go look).

Reply:
I see your point and agree. Let's wait to see the approach and together with that I'll modify this as well.

Reply:
@gaborgsomogyi , me confused , how to overwrite this Utils.createTempDir in spark-streaming code which uses spark-sql.2.4.1 version ?